### PR TITLE
feat: API explorer error boundary

### PR DESCRIPTION
**Pull request by bryans99**
# Description

Adds an error boundary around APIX explorer child components to display a "nice" message should an error occur during component renders (note, error boundaries do not help with asynchronous errors).

# Changes

1. The error boundary and its child components.
2. Added support to environment adapter for logging. Only extension api explorer logs at the moment as logging strategy for stand alone apix explorer is TBD.
3, Added ability to use svgs in react code. Had to add `custom.d.ts` file to api-extension and extension-api-extension projects for `tsc` compiler to work. There may be better ways to do this but that can be added later.

# Screen shots

## Stand alone (note no header)

image

## Extension - header is owned by looker host

image

### CODE REVIEW

File: packages/api-explorer/src/ApiExplorer.tsx [hunk 1]
```diff
@@ -125,42 +124,44 @@ const ApiExplorer: FC<ApiExplorerProps> = ({
...
-                  {hasNavigation && (
-                    <AsideBorder pt="large" width="20rem">
-                      <SideNav
-                        headless={headless}
-                        specs={specs}
-                        spec={spec}
-                        specDispatch={specDispatch}
-                      />
-                    </AsideBorder>
+        <ErrorBoundary logError={envAdaptor.logError.bind(envAdaptor)}>
```
Comment (jkaster) [inline packages/api-explorer/src/ApiExplorer.tsx:127]: This level certainly captures error handling!

File: packages/api-explorer/src/components/ErrorBoundary/ErrorBoundary.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    const badObj: any = {}
+    return badObj.a.b
+  }
+  return <div>{text + terminateWithPrejudice()}</div>
+}
+
+describe('ErrorBoundary', () => {
+  const header = 'Uh oh! Something went wrong!'
+  const actionMessage = 'Try refreshing the page the correct the problem.'
+  const childText = 'Sweet Child'
```
Comment (jkaster) [inline packages/api-explorer/src/components/ErrorBoundary/ErrorBoundary.spec.tsx:41]: of mine

File: packages/api-explorer/src/components/ErrorBoundary/ErrorBoundary.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  })
+
+  it('should render error detail on error', () => {
+    const logError = jest.fn()
+    const { getByText, queryByText } = renderWithTheme(
+      <ErrorBoundary logError={logError}>
+        <ErrorComp text={childText} />
+      </ErrorBoundary>
+    )
+    expect(queryByText(childText)).not.toBeInTheDocument()
```
Comment (jkaster) [inline packages/api-explorer/src/components/ErrorBoundary/ErrorBoundary.spec.tsx:67]: Why `.not.`? Is it shown in "something went wrong" instead?

File: packages/api-explorer/src/custom.d.ts [hunk 4]
```diff
@@ -0,0 +1,4 @@
+declare module '*.svg' {
+  const content: any
+  export default content
+}
```
Comment (jkaster) [inline packages/api-explorer/src/custom.d.ts:4]: interesting "subscribe"

File: packages/api-explorer/src/utils/envAdaptor.ts [hunk 5]
```diff
@@ -79,7 +81,12 @@ export class StandaloneEnvAdaptor implements IApixEnvAdaptor {
     this._themeOverrides = getThemeOverrides(
       hostname.endsWith('.looker.com') ||
         hostname.endsWith('.google.com') ||
-        hostname === 'localhost'
+        hostname === 'localhost' ||
+        // Include firebase staging dev portal for now. Can be removed
+        // when dev portal gets its own APIX project. Also includes
+        // PRs.
+        (hostname.startsWith('looker-developer-portal') &&
+          hostname.endsWith('.web.app'))
```
Comment (jkaster) [inline packages/api-explorer/src/utils/envAdaptor.ts:89]: Thanks for the explanation

File: packages/extension-api-explorer/src/custom.d.ts [hunk 6]
```diff
@@ -0,0 +1,4 @@
+declare module '*.svg' {
+  const content: any
+  export default content
+}
```
Comment (jkaster) [inline packages/extension-api-explorer/src/custom.d.ts:4]: Interesting we need this in both api-explorer and here

File: webpack.base.config.js [hunk 7]
```diff
@@ -47,6 +47,15 @@ module.exports = function(packagePath = __dirname) {
           test: /\.css$/i,
           use: ['style-loader', 'css-loader'],
         },
+        {
+          test: /\.(png|woff(2)?|eot|ttf|svg)$/,
+          loader: 'url-loader',
```
Comment (jkaster) [inline webpack.base.config.js:52]: Nice to see how to load assets

**Review by jkaster — APPROVED** [review]
I think this is ok. Just have a clarification question.

File: packages/api-explorer/src/components/ErrorBoundary/ErrorBoundary.spec.tsx (hunk 2 above)
Comment (bryans99) [reply packages/api-explorer/src/components/ErrorBoundary/ErrorBoundary.spec.tsx:41]: That was the song that was going through my head when I wrote the test :D

File: packages/api-explorer/src/components/ErrorBoundary/ErrorBoundary.spec.tsx (hunk 3 above)
Comment (bryans99) [reply packages/api-explorer/src/components/ErrorBoundary/ErrorBoundary.spec.tsx:67]: It either shows the error text OR the children. Presumably something in the children cause the error and for the error boundary to be notified.

File: packages/extension-api-explorer/src/custom.d.ts (hunk 6 above)
Comment (bryans99) [reply packages/extension-api-explorer/src/custom.d.ts:4]: Yeah. I alluded to this in the commit message. I tried various ways tsconfig options and putting the file in a single location but did not have any luck. This is the only way I got the tsc command to work. babel does not have any issues but we are still using tsc for emitting the type declarations

File: webpack.base.config.js (hunk 7 above)
Comment (bryans99) [reply webpack.base.config.js:52]: Stolen from HT :D

**Pull request merged**